// @azure/event-hubs/dist-esm/src/pumpManager.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { __awaiter } from "tslib";
import { CloseReason } from "./models/public";
import { PartitionPump } from "./partitionPump";
import { logger, logErrorStackTrace } from "./log";
/**
 * The PumpManager handles the creation and removal of PartitionPumps.
 * It also starts a PartitionPump when it is created, and stops a
 * PartitionPump when it is removed.
 * @ignore
 * @internal
 */
export class PumpManagerImpl {
    /**
     * @ignore
     */
    constructor(eventProcessorName, eventProcessorOptions) {
        this._partitionIdToPumps = {};
        this._eventProcessorName = eventProcessorName;
        this._options = eventProcessorOptions;
    }
    /**
     * Returns a list of partitionIds that are actively receiving messages.
     * @ignore
     */
    receivingFromPartitions() {
        return Object.keys(this._partitionIdToPumps).filter((id) => {
            const pump = this._partitionIdToPumps[id];
            return Boolean(pump && pump.isReceiving);
        });
    }
    /**
     * Indicates whether the pump manager is actively receiving events from a given partition.
     * @param partitionId
     * @ignore
     * @internal
     */
    isReceivingFromPartition(partitionId) {
        const pump = this._partitionIdToPumps[partitionId];
        return Boolean(pump && pump.isReceiving);
    }
    /**
     * Creates and starts a PartitionPump.
     * @param startPosition The position in the partition to start reading from.
     * @param eventHubClient The EventHubClient to forward to the PartitionPump.
     * @param partitionProcessor The PartitionProcessor to forward to the PartitionPump.
     * @ignore
     */
    createPump(startPosition, eventHubClient, partitionProcessor, abortSignal) {
        return __awaiter(this, void 0, void 0, function* () {
            const partitionId = partitionProcessor.partitionId;
            if (abortSignal.aborted) {
                logger.verbose(`${this._eventProcessorName}] The subscription was closed before creating the pump for partition ${partitionId}.`);
                return;
            }
            // attempt to get an existing pump
            const existingPump = this._partitionIdToPumps[partitionId];
            if (existingPump) {
                if (existingPump.isReceiving) {
                    logger.verbose(`[${this._eventProcessorName}] [${partitionId}] The existing pump is running.`);
                    return;
                }
                logger.verbose(`[${this._eventProcessorName}] [${partitionId}] The existing pump is not running.`);
                yield this.removePump(partitionId, CloseReason.OwnershipLost);
            }
            logger.verbose(`[${this._eventProcessorName}] [${partitionId}] Creating a new pump.`);
            const pump = new PartitionPump(eventHubClient, partitionProcessor, startPosition, this._options);
            try {
                yield pump.start();
                this._partitionIdToPumps[partitionId] = pump;
            }
            catch (err) {
                logger.verbose(`[${this._eventProcessorName}] [${partitionId}] An error occured while adding/updating a pump: ${err}`);
                logErrorStackTrace(err);
            }
        });
    }
    /**
     * Stop a PartitionPump and removes it from the internal map.
     * @param partitionId The partitionId to remove the associated PartitionPump from.
     * @param reason The reason for removing the pump.
     * @ignore
     */
    removePump(partitionId, reason) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const pump = this._partitionIdToPumps[partitionId];
                if (pump) {
                    delete this._partitionIdToPumps[partitionId];
                    logger.verbose(`[${this._eventProcessorName}] [${partitionId}] Stopping the pump.`);
                    yield pump.stop(reason);
                }
                else {
                    logger.verbose(`[${this._eventProcessorName}] [${partitionId}] No pump was found to remove.`);
                }
            }
            catch (err) {
                logger.verbose(`[${this._eventProcessorName}] [${partitionId}] An error occured while removing a pump: ${err}`);
                logErrorStackTrace(err);
            }
        });
    }
    /**
     * Stops all PartitionPumps and removes them from the internal map.
     * @param reason The reason for removing the pump.
     * @ignore
     */
    removeAllPumps(reason) {
        return __awaiter(this, void 0, void 0, function* () {
            const partitionIds = Object.keys(this._partitionIdToPumps);
            logger.verbose(`[${this._eventProcessorName}] Removing all pumps due to reason ${reason}.`);
            const tasks = [];
            for (const partitionId of partitionIds) {
                const pump = this._partitionIdToPumps[partitionId];
                if (pump) {
                    tasks.push(pump.stop(reason));
                }
            }
            try {
                yield Promise.all(tasks);
            }
            catch (err) {
                logger.verbose(`[${this._eventProcessorName}] An error occured while removing all pumps: ${err}`);
                logErrorStackTrace(err);
            }
            finally {
                this._partitionIdToPumps = {};
            }
        });
    }
}
